### DISCUSSION TITLE
RPC requests

### DISCUSSION THREAD

**Discussion by leopashov** [MetaMask/snaps#1851] [Q&A] [answered]
I have seen the possibility of getting snaps to communicate with each other (https://docs.metamask.io/snaps/reference/exports/#onrpcrequest), but can't work out how to actually implement this functionality.
Are there any example projects which are making use of this? Or any code snippets one could direct me to?
Initially I'm just looking to create a simple example where I could pass some data from one snap to another.
So far I have added some custom functionality (mostly generating keys) to the snaps monorepo project, and have seen bits about the keyring handler and custom handler at https://docs.metamask.io/snaps/tutorials/custom-evm-accounts/#2-expose-the-keyring-interface-as-a-json-rpc-api, but can't figure out how to go on to use these.

I would greatly appreciate any help or guidance :)

**Comment by Montoya** [comment, marked as answer]
There are no good examples for this because the UX doesn't really support this yet. The way it has to be done is the user has to install both Snaps, then the caller Snap has to "connect" to the responder Snap. And `endowment:rpc` has to have the `snaps` caveat set to `true`.

**Reply by leopashov** [reply]
Thank you very much for your response, glad to hear there is a valid reason for my struggles!
So far I have only had one snap installed at a time, is there any guidance on how one might install two snaps?
In terms of 'connecting' to the responder snap, would this then have to be done via a dapp, or the separate snaps (or instances of the snaps) could communicate through metamask? 
Is there any code of this being implemented - maybe even in tests - which could help my understanding?
Thanks again :)

**Reply by leopashov** [reply]
@Montoya Could I ask what you mean with regard to the caller Snap 'connecting' to the responder snap?
Initially I thought this may just be a call from the dapp, but realise this would not actually be snap to snap communication.
How would the snap be called from within Metamask? Is there a way of getting a list of installed snaps, for example, and then being able to access the methods within these? Or is this the functionality you mention hasn't been built yet?

**Reply by aquiladev** [reply]
We figured out with @ziad-saab that it is not possible to connect snap to another snap, due to blocked RPC method. We are trying to overcome it...

**Reply by FrederikBolding** [reply]
@aquiladev This is not entirely accurate. You can connect to another snap but you cannot do it dynamically. The permission has to be requested like so: https://github.com/MetaMask/snaps/blob/main/packages/examples/packages/json-rpc/snap.manifest.json#L23-L27

And the other snap has to be installed, then you can invoke it.

**Reply by aquiladev** [reply]
Still does not work. Target snap is installed
```
"initialPermissions": {
    "endowment:rpc": {
      "dapps": true,
      "snaps": true
    },
    "wallet_snap": {
      "npm:@consensys/starknet-snap": {}
    }
  },
```

Code:
```
const res = await snap.request({
        method: 'wallet_invokeSnap',
        params: {
          snapId: 'npm:@consensys/starknet-snap',
          request: { method: 'starkNet_recoverAccounts' },
        },
 });
 console.log(res);
```

Error:
```
"Snap "npm:@consensys/starknet-snap" is not permitted to handle requests from "local:http://localhost:8080"."
```

**Reply by ziad-saab** [reply]
@aquiladev The StarkNet Snap doesn't allow RPC calls from other Snaps unfortunately:

https://github.com/Consensys/starknet-snap/blob/main/packages/starknet-snap/snap.manifest.json#L30

**Reply by aquiladev** [reply]
Indeed, static configuration works well. The only problem that target snaps must allow it via 'snaps: true'. Thanks for help!

**Reply by leopashov** [reply]
Thanks very much all, making more sense now :)
